lang/java/reef-io/src/main/java/org/apache/reef/io/network/group/impl/operators/GatherReceiver.java

```diff
@@ -0,0 +1,171 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class GatherReceiver<T> implements Gather.Receiver<T>, EventHandler<GroupCommunicationMessage> {
+
+  private static final Logger LOG = Logger.getLogger(GatherReceiver.class.getName());
+
+  private final Class<? extends Name<String>> groupName;
+  private final Class<? extends Name<String>> operName;
+  private final CommGroupNetworkHandler commGroupNetworkHandler;
+  private final Codec<T> dataCodec;
+  private final NetworkService<GroupCommunicationMessage> netService;
+  private final Sender sender;
+
+  private final OperatorTopology topology;
+
+  private final CommunicationGroupServiceClient commGroupClient;
+
+  private final AtomicBoolean init = new AtomicBoolean(false);
+
+  private final int version;
+
+  @Inject
+  public GatherReceiver(@Parameter(CommunicationGroupName.class) final String groupName,
+                        @Parameter(OperatorName.class) final String operName,
+                        @Parameter(TaskConfigurationOptions.Identifier.class) final String selfId,
+                        @Parameter(DataCodec.class) final Codec<T> dataCodec,
+                        @Parameter(DriverIdentifier.class) final String driverId,
+                        @Parameter(TaskVersion.class) final int version,
+                        final CommGroupNetworkHandler commGroupNetworkHandler,
+                        final NetworkService<GroupCommunicationMessage> netService,
+                        final CommunicationGroupServiceClient commGroupClient) {
+    super();
+    LOG.finest(operName + " has CommGroupHandler-" + commGroupNetworkHandler.toString());
+    this.version = version;
+    this.groupName = Utils.getClass(groupName);
+    this.operName = Utils.getClass(operName);
+    this.dataCodec = dataCodec;
+    this.commGroupNetworkHandler = commGroupNetworkHandler;
+    this.netService = netService;
+    this.sender = new Sender(this.netService);
+    this.topology = new OperatorTopologyImpl(this.groupName, this.operName, selfId, driverId, sender, version);
+    this.commGroupNetworkHandler.register(this.operName, this);
+    this.commGroupClient = commGroupClient;
+  }
+
+  @Override
+  public int getVersion() {
+    return version;
+  }
+
+  @Override
+  public void initialize() throws ParentDeadException {
+    topology.initialize();
+  }
+
+  @Override
+  public Class<? extends Name<String>> getOperName() {
+    return operName;
+  }
+
+  @Override
+  public Class<? extends Name<String>> getGroupName() {
+    return groupName;
+  }
+
+  @Override
+  public String toString() {
+    return "GatherReceiver:" + Utils.simpleName(groupName) + ":" + Utils.simpleName(operName) + ":" + version;
+  }
+
+  @Override
+  public void onNext(final GroupCommunicationMessage msg) {
+    topology.handle(msg);
+  }
+
+  @Override
+  public List<T> receive() throws NetworkException, InterruptedException {
```

Comment:
Sorry about that, there weren't any comments in the `Gather` interface that mentions lexicographical order requirements. I can refactor the code to return elements in lexicographically sorted id order, but before I do I'd like to politely ask why we'd want to implement `Gather` that way. In an elastic environment, the task identifiers wouldn't matter much, would they? Please correct me if I'm wrong.